### PR TITLE
Placeholder search

### LINKED ISSUE #729: Tracking issue: Placeholder search or a search without query parameters
This feature will be helpful for ecommerce websites or any kind of website that needs to display documents without specifying any search query. The displayed documents are sorted by the custom criteria specified in the settings when the query parameter is empty or not present at all (i.e `/search?q=` or `/search`).

[The `sortBy` feature](https://github.com/meilisearch/MeiliSearch/issues/730) will not be supported by placeholders, but we will eventually support it.

### What do we do at indexing time?

  1. Retrieve the custom ranking criteria in the order specified in the settings.
      If the settings contains `typo, words, asc(price), desc(time), proximity` the custom criteria that we will use are `asc(price)` and `desc(time)` in this specific order.
  2. Sort all the documents ids by the custom criteria.
  3. Store those ids in the key-value store ~and keep the deepest buckets bounds~.

### At query time with default search settings

We must use the cached documents ids and return the documents lying between the bounds requested (the result page).

### At query time with facets filters

  1. We compute the intersections and unions of the facets filters documents ids.
  2. Compute the scores of this restricted list of ids using the customs criteria defined in the settings.

### PR DESCRIPTION
**Pull request by MarinPostma**
close #729

### CODE REVIEW

File: meilisearch-core/src/query_builder.rs [hunk 1]
```diff
@@ -78,67 +74,73 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
     }
 
     pub fn add_searchable_attribute(&mut self, attribute: u16) {
-        let reorders = self.searchable_attrs.get_or_insert_with(ReorderedAttrs::new);
+        let reorders = self
+            .searchable_attrs
+            .get_or_insert_with(ReorderedAttrs::new);
```
Comment (MarinPostma) [inline meilisearch-core/src/query_builder.rs:79]:
```suggestion
        let reorders = self.searchable_attrs.get_or_insert_with(ReorderedAttrs::new);
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 2]
```diff
@@ -145,6 +145,16 @@ where A: AsRef<[u8]>,
     Ok(())
 }
 
+fn cache_documents_document_ids_sorted<'a>(
+    writer: &'a mut heed::RwTxn<MainT>,
+    ranked_map: &RankedMap,
+    index: &store::Index,
+    mut document_ids: Vec<DocumentId>,
+) -> MResult<()> {
+    crate::bucket_sort::placeholder_document_sort(&mut document_ids, index, writer, ranked_map)?;
```
Comment (MarinPostma) [inline meilisearch-core/src/update/documents_addition.rs:154]:
```suggestion
    document_ids: &mut [DocumentId],
) -> MResult<()> {
    crate::bucket_sort::placeholder_document_sort(document_ids, index, writer, ranked_map)?;
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 3]
```diff
@@ -233,12 +233,13 @@ pub fn apply_addition<'a, 'b>(
...
                 &schema,
                 field_id,
-                document_id,
+                *document_id,
                 &value,
             )?;
         }
     }
 
+
```
Comment (MarinPostma) [inline meilisearch-core/src/update/documents_addition.rs:242]:
```suggestion
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 4]
```diff
@@ -327,12 +332,13 @@ pub fn reindex_all_documents(writer: &mut heed::RwTxn<MainT>, index: &store::Ind
...
```
Comment (MarinPostma) [inline meilisearch-core/src/update/documents_addition.rs:341]:
```suggestion
```

File: meilisearch-core/src/query_builder.rs [hunk 5]
```diff
@@ -82,63 +78,67 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
...
 
-    pub fn query(
-        self,
-        reader: &heed::RoTxn<MainT>,
-        query: &str,
-        range: Range<usize>,
-    ) -> MResult<SortResult> {
-        let facets_docids = match self.facet_filter {
-            Some(facets) => {
+    /// returns the documents ids ascociated with a facet filter by computing the union and
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:81]:
```suggestion
    /// returns the documents ids associated with a facet filter by computing the union and
```

File: meilisearch-core/src/query_builder.rs [hunk 6]
```diff
@@ -82,63 +78,67 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
...
         };
+        Ok(facet_docids)
+    }
 
+    fn standard_query(
+        self,
+        reader: &MainReader,
+        query: &str,
+        range: Range<usize>,
+    ) -> MResult<SortResult> {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:133]:
```suggestion
    fn standard_query(self, reader: &MainReader, query: &str, range: Range<usize>) -> MResult<SortResult> {
```

File: meilisearch-core/src/query_builder.rs [hunk 7]
```diff
@@ -177,6 +177,124 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
             ),
         }
     }
+
+    fn placeholder_query(
+        self,
+        reader: &heed::RoTxn<MainT>,
+        range: Range<usize>,
+    ) -> MResult<SortResult> {
```
Comment (Kerollmops) [inline meilisearch-core/src/query_builder.rs:185]:
```suggestion
    fn placeholder_query(self, reader: &heed::RoTxn<MainT>, range: Range<usize>) -> MResult<SortResult> {
```

File: meilisearch-core/src/store/main.rs [hunk 8]
```diff
@@ -29,6 +29,7 @@ const STOP_WORDS_KEY: &str = "stop-words";
 const SYNONYMS_KEY: &str = "synonyms";
 const UPDATED_AT_KEY: &str = "updated-at";
 const WORDS_KEY: &str = "words";
+const SORTED_DOCUMENT_IDS_CACHE: &str = "sorted-document-ids-cache";
```
Comment (Kerollmops) [inline meilisearch-core/src/store/main.rs:32]:
```suggestion
const SORTED_DOCUMENT_IDS_CACHE_KEY: &str = "sorted-document-ids-cache";
```

Comment (Kerollmops) [reply meilisearch-core/src/store/main.rs:32]: Please sort it among the others :)

File: meilisearch-core/src/update/documents_deletion.rs [hunk 9]
```diff
@@ -180,5 +180,9 @@ pub fn apply_documents_deletion(
 
     compute_short_prefixes(writer, &words, index)?;
 
+    // update is finished; update sorted document id cache with new state
+    let mut document_ids = index.main.internal_docids(writer)?.to_vec();
+    super::cache_documents_document_ids_sorted(writer, &ranked_map, index, &mut document_ids)?;
```
Comment (Kerollmops) [inline meilisearch-core/src/update/documents_deletion.rs:185]: Should the `document_ids` here be the whole list of all documents ids of the database?
If so, are you sure this is the case? Isn't this list only the documents ids related to this specific update?

File: meilisearch-http/tests/placeholder_search.rs [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+        let get_query: SearchQuery = post_query.into();
+        let get_query = ::serde_url_params::to_string(&get_query).unwrap();
+        let ($response, $status_code) = $server.search_get(&get_query).await;
+        let _ =::std::panic::catch_unwind(|| $block)
+            .map_err(|e| panic!("panic in get route: {:?}", e.downcast_ref::<&str>().unwrap()));
+        let ($response, $status_code) = $server.search_post($query).await;
+        let _ = ::std::panic::catch_unwind(|| $block)
+            .map_err(|e| panic!("panic in post route: {:?}", e.downcast_ref::<&str>().unwrap()));
+    };
+}
```
Comment (Kerollmops) [inline meilisearch-http/tests/placeholder_search.rs:23]: Why don't you use the other macro you've made?

File: meilisearch-core/src/update/documents_deletion.rs (hunk 9 above)
Comment (MarinPostma) [reply meilisearch-core/src/update/documents_deletion.rs:185]: there may be documents in the database that are not in the update right? I need to sort them too with the new documents

File: meilisearch-http/tests/placeholder_search.rs (hunk 10 above)
Comment (MarinPostma) [reply meilisearch-http/tests/placeholder_search.rs:23]: I would have to expose the search.rs module to the placeholder_search module, or move this macro to common, where it doesn't really belong, this seemed to be the cleanest option, wdyt?

**Comment by curquiza** [comment]
Can you add the label `meilisearch-http` since only this PR is added to the v0.13.0 Milestones and no the issue.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,8 @@
+## v0.12.0 (unreleased)
```
Comment (Kerollmops) [inline CHANGELOG.md:1]:
```suggestion
## v0.13.0 (unreleased)
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 12]
```diff
@@ -255,6 +255,10 @@ pub fn apply_addition<'a, 'b>(
         let facet_map = facets::facet_map_from_docids(writer, index, &docids, attributes_for_facetting.as_ref())?;
         index.facets.add(writer, facet_map)?;
     }
+    //
```
Comment (Kerollmops) [inline meilisearch-core/src/update/documents_addition.rs:258]:
```suggestion
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 13]
```diff
@@ -352,6 +356,10 @@ pub fn reindex_all_documents(writer: &mut heed::RwTxn<MainT>, index: &store::Ind
...
```
Comment (Kerollmops) [inline meilisearch-core/src/update/documents_addition.rs:359]:
```suggestion
```

File: meilisearch-http/tests/common.rs [hunk 14]
```diff
@@ -1,15 +1,31 @@
...
+        let get_query: meilisearch_http::routes::search::SearchQuery = post_query.into();
+        let get_query = ::serde_url_params::to_string(&get_query).unwrap();
+        let ($response, $status_code) = $server.search_get(&get_query).await;
+        let _ =::std::panic::catch_unwind(|| $block)
+            .map_err(|e| panic!("panic in get route: {:?}", e.downcast_ref::<&str>().unwrap()));
+        let ($response, $status_code) = $server.search_post($query).await;
+        let _ = ::std::panic::catch_unwind(|| $block)
+            .map_err(|e| panic!("panic in post route: {:?}", e.downcast_ref::<&str>().unwrap()));
+    };
+}
```
Comment (Kerollmops) [inline meilisearch-http/tests/common.rs:28]: Why do I see this macro in this PR?

Comment (MarinPostma) [reply meilisearch-http/tests/common.rs:28]: This was moved to `common` to be shared with the placeholder search tests :)

**Pull request merged**